src/redjubjub.rs

```diff
@@ -0,0 +1,137 @@
+//! Implementation of RedJubjub, a specialization of RedDSA to the Jubjub curve.
+//! See section 5.4.6 of the Sapling protocol specification.
+
+use pairing::Field;
+use rand::Rng;
+
+use jubjub::{FixedGenerators, JubjubEngine, JubjubParams, Unknown, edwards::Point};
+use util::hash_to_scalar;
+
+fn h_star<E: JubjubEngine>(a: &[u8], b: &[u8]) -> E::Fs {
+    hash_to_scalar::<E>(b"Zcash_RedJubjubH", a, b)
+}
+
+pub struct Signature<E: JubjubEngine> {
+    r: Point<E, Unknown>,
+    s: E::Fs,
+}
+
+pub struct PrivateKey<E: JubjubEngine>(E::Fs);
+
+pub struct PublicKey<E: JubjubEngine>(Point<E, Unknown>);
+
+impl<E: JubjubEngine> PrivateKey<E> {
+    pub fn randomize(&self, alpha: E::Fs) -> Self {
+        let mut tmp = self.0;
+        tmp.add_assign(&alpha);
+        PrivateKey(tmp)
+    }
+
+    pub fn sign<R: Rng>(&self, msg: &[u8], rng: &mut R, params: &E::Params) -> Signature<E> {
+        // T = (l_H + 128) bits of randomness
+        // For H*, l_H = 512 bits
+        let mut t = [0u8; 80];
+        rng.fill_bytes(&mut t[..]);
+
+        // r = H*(T || M)
+        let r = h_star::<E>(&t[..], msg);
+
+        // R = r . G
+        let r_g = params
+            .generator(FixedGenerators::SpendingKeyGenerator)
+            .mul(r, params);
+        let mut rbar = [0u8; 32];
+        r_g.write(&mut rbar[..])
+            .expect("Jubjub points should serialize to 32 bytes");
+
+        // S = r + H*(Rbar || M) . sk
+        let mut s = h_star::<E>(&rbar[..], msg);
+        s.mul_assign(&self.0);
+        s.add_assign(&r);
+
+        Signature { r: r_g.into(), s }
+    }
+}
+
+impl<E: JubjubEngine> PublicKey<E> {
+    pub fn from_private(privkey: &PrivateKey<E>, params: &E::Params) -> Self {
+        let res = params
+            .generator(FixedGenerators::SpendingKeyGenerator)
+            .mul(privkey.0, params)
+            .into();
+        PublicKey(res)
+    }
+
+    pub fn randomize(&self, alpha: E::Fs, params: &E::Params) -> Self {
+        let res: Point<E, Unknown> = params
+            .generator(FixedGenerators::SpendingKeyGenerator)
+            .mul(alpha, params)
+            .into();
+        let res = res.add(&self.0, params);
+        PublicKey(res)
+    }
+
+    // Pre-conditions:
+    // - rbar was the canonical representation of a point on the curve.
+    // - sig.s < order(G)
+    // TODO(str4d): Enforce these during deserialization of Signature
+    pub fn verify(&self, msg: &[u8], sig: &Signature<E>, params: &E::Params) -> bool {
+        // c = H*(Rbar || M)
+        let mut rbar = [0u8; 32];
+        sig.r
+            .write(&mut rbar[..])
+            .expect("Jubjub points should serialize to 32 bytes");
```

Comment:
If I'm reading correctly, we're decoding a point representation (outside this class), reencoding it here, and relying on the fact that other code enforced a canonical representation. That seems ugly and error-prone.

Reply:
This is an artifact of @ebfull's desire to represent `Signature` in-memory as the unserialized values. We could alternatively represent it as the serialized string, and unserialize it every time we verify the signature. Or we could store both `Rbar` and `R` in the signature, which I don't like because it feels like it could get out-of-sync.

Reply:
It was an intentional part of the design (inherited from EdDSA) that you hash the representation of R exactly as given in the signature. That implies that any change to the representation R̲ will result in a completely different hash, and so it's not possible to malleate the signature by changing R̲. (It still might be by changing S̲, but that's a simpler encoding.)

I realize it is equivalent if the representation of R is ensured to be canonical, but then you have two places to look in the code to check that the signature is being verified correctly. As specified, a RedJubjub signature is a byte sequence, *not* a (point, integer) pair, and so I recommend that it be implemented that way for ease of auditing.